doc/sg_rem_rest_elem.8
.TH SG_REM_REST_ELEM "8" "May 2023" "sg3_utils\-1.48" SG3_UTILS
.SH NAME
sg_rem_rest_elem \- send SCSI remove or restore element command
.SH SYNOPSIS
.B sg_rem_rest_elem
[\fI\-\-capacity=RC\fR] [\fI\-\-element=EID\fR] [\fI\-\-help\fR]
[\fI\-\-quick\fR] [\fI\-\-remove\fR] [\fI\-\-restore\fR]
[\fI\-\-timeout=SE\fR] [\fI\-\-verbose\fR] [\fI\-\-version\fR]
\fIDEVICE\fR
.SH DESCRIPTION
.\" Add any additional description here
Sends a SCSI REMOVE ELEMENT AND TRUNCATE [RMEAT] or RESTORE ELEMENTS AND
REBUILD [RSEAR] command to the \fIDEVICE\fR. Since both these commands have
a potentially huge impact on the \fIDEVICE\fR (similar to the FORMAT UNIT
command: destroying data and taking a long time to complete fully),
they first give the user the chance to reconsider (3 times within 15
seconds) before taking action.
.PP
Unlike the FORMAT UNIT command, these commands seem designed to work in
the background. So they will return quickly (although sbc5r01.pdf does not
state that) and the disk will be placed in a reduced functionality state
where only a specified number of commands will be executed (e.g. INQUIRY and
REPORT LUNS) until the operation is complete. Other commands will receive
sense data with a sense key of NOT READY and an additional sense code
of 'Depopulation in progress' (for RMEAT) or 'Depopulation restoration in
progress' (for RSEAR).
.PP
The REMOVE ELEMENT AND TRUNCATE has a close relative in ZBC\-2 called the
REMOVE ELEMENT AND MODIFY ZONES [RMEMZ] command. See the sg_zone utility
for an implementation of the latter command.
.br
The difference between RMEAT and RMEMZ is that the former "changes the
association between LBAs and physical blocks" and the latter does not
change that association. Zones affected by the RMEMZ command are placed
into the zone condition: "Offline".
.SH OPTIONS
Arguments to long options are mandatory for short options as well.
.TP
\fB\-c\fR, \fB\-\-capacity\fR=\fIRC\fR
RC stands for Requested Capacity and is the number of logical blocks the
\fIDEVICE\fR should have after the element is removed with the RMEAT
command. The default value is 0 which allows the \fIDEVICE\fR to decide
what the reduced capacity will be after the element removal. The RSEAR
command ignores this value.
.TP
\fB\-e\fR, \fB\-\-element\fR=\fIEID\fR
where \fIEID\fR is an element identifier which is a 32 bit unsigned integer
starting at one. This field is used by the RMEAT command and ignored
otherwise. The default value is zero (which is invalid). So the user needs
to supply a valid element identifier when \fI\-\-remove\fR is used.
.TP
\fB\-h\fR, \fB\-\-help\fR
output the usage message then exit.
.TP
\fB\-q\fR, \fB\-\-quick\fR
the default action (i.e. when this option is not given) is to give the user
15 seconds to reconsider doing a remove or restore element operation on the
\fIDEVICE\fR.  When this option is given that step (i.e. the 15 second
warning period) is bypassed.
.TP
\fB\-r\fR, \fB\-\-remove\fR
causes the REMOVE ELEMENT AND TRUNCATE command to be sent to the
\fIDEVICE\fR. In practice, \fI\-\-element=EID\fR needs to be also given.
.TP
\fB\-R\fR, \fB\-\-restore\fR
causes the RESTORE ELEMENTS AND REBUILD command to be sent to the
\fIDEVICE\fR.
.TP
\fB\-t\fR, \fB\-\-timeout\fR=\fISE\fR
where \fISE\fR is the command timeout in seconds. The default is 60 seconds
and if 0 is given, it is mapped to 60. An alternate long form is
\fI\-\-tmo=SE\fR.
.TP
\fB\-v\fR, \fB\-\-verbose\fR
increase the level of verbosity, (i.e. debug output).
.TP
\fB\-V\fR, \fB\-\-version\fR
print the version string and then exit.
.SH NOTES
Once an element is removed successfully it is termed as "depopulated".
Depopulated elements that have the 'Restoration Allowed' (RALWD) bit
set (see sg_get_elem_status) are candidates for future restoration.
.PP
A (storage) element of a rotating hard disk is one side of a platter
typically associated with one head. Such hard disks typically have multiple
platters with two heads per platter (i.e. one head each side of the platter).
.SH EXIT STATUS
The exit status of sg_rem_rest_elem is 0 when it is successful. Otherwise see
the sg3_utils(8) man page.
.SH AUTHORS
Written by Douglas Gilbert.
.SH "REPORTING BUGS"
Report bugs to <dgilbert at interlog dot com>.
.SH COPYRIGHT
Copyright \(co 2022\-2023 Douglas Gilbert
.br
This software is distributed under a BSD\-2\-Clause license. There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.SH "SEE ALSO"
.B sg_get_elem_status,sg_zone(sg3_utils)


CHANGELOG
Each utility has its own version number, date of last change and
some description at the top of its ".c" file. All utilities in the main
directory have their own "man" pages. There is also a sg3_utils man page.

Changelog for sg3_utils-1.25 [20070710]
  - sg_dd: add oflag=sparse to step over bs*bpt number of zeros;
    - with oflag=sparse, write last bs*bpt segment at end or after
      error so file length of OFILE is appropriate
    - allow iflag=null and oflag=null which do nothing (placeholders)
    - when coe>1 then SCSI READ LONG logic remember extended block
      length of first encountered error
  - sg_ses: sync with ses2r17
  - place source in subversion repository

Changelog for sg3_utils-1.24 [20070507]
  - sg_raw: new utility to send arbirary SCSI commands
  - sg_luns: increase number of luns that can be fetched
    - fix length of raw and hex output
    - add '--quiet' option to output only ASCII hex
      representation of each lun
  - sg_rtpg: update for changes in spc4r09
  - sg_persist: update documentation, spc-4 references
    - fix exit status values
  - sg_inq: update version descriptors per spc4r09
    - fix '--id' and '--extended'
    - extend block limits VPD page (sbc3r09)
  - sg_vpd: extend block limits VPD page (sbc3r09)
    - append relative target port identifier to SAS target
      port address with '-iq' option
  - sg_logs: add decoding for stats+performance log pages
    - fix showing of page names for pdt > 0
    - implement '-HH' for single and all pages, fix '-r'
    - when '--maxlen=' given, only do single fetch
    - add Tape Alert (ssc), Media and Element statistics (smc) pages 
    - add '--brief' option
  - sg_ses: sync with ses2r16
    - fix bay number for SAS
  - sg_format: add '--dcrt' and '--security' options
  - sgm_dd: add 'smmap' oflag for shared_mmap_io testing
    - add 'dio' oflag
  - sg_dd, sgp_dd: add 'dio' iflag and oflag
  - sg_modes: change SAS mode page names per sas2r09
    - check validity of block descriptors length
  - sg_pt: change opaque context object from 'void *'
    to 'struct sg_pt_base *'
  - sg_opcodes: anticipate extra tmfs from 07-159r0
  - sg_sat_set_features: add more usage information
    - add man page
  - sg_sat_phy_event: add to examples directory
  - sg_lib: sync asc/ascq codes with spc4r10
  - Solaris port: using uscsi interface

Changelog for sg3_utils-1.23 [20070131]
  - sg_read_buffer: new utility
  - sg_write_buffer: new utility
  - sg_opcodes, sg_senddiag, sg_logs, sg_modes, sg_start, sg_inq,
    sg_turs, sg_readcap, sg_rbuf: add getopt_long() based cli;
    old and new cli selectable, new getopt_long cli is default
  - scripts: new subdirectory containing some bash scripts
    - add scripts/README file
  - sg_reassign: add '--hex' option for grown and primary lists
  - sg_rtpg: add '--raw' option
  - sg_lib.h, sg_cmds_basic.h + sg_cmds_extra.h: add C++ 
    'extern "C" ' wrappers
    - cleanup C code so it will compile as C++
  - sg_lib: sync with spc4r08
    - include <inttypes.h>, use PRId64 instead of %lld form
    - fix sg_get_sense_str() when empty sense buffer
  - win32 port: add Makefile.mingw + related support for MinGW
  - sg_cmds_extra: add sg_ll_read_buffer() and sg_ll_write_buffer()
  - sg_dd, sgp_dd, sgm_dd, sg_read: use lseek64() instead of llseek.c
  - sgm_dd: accept coe=<n> for interworking with sg_dd
  - sg_rdac: fix on non-linux ports
  - sg_ses: fix spurious warning in additional element status page
    - '-rr' option outputs a diagnostic page in binary to stdout
  - sg_opcodes: add command timeout descriptor support (spc4r08)
    - change linux specific pass through to generic pass through
  - sg_logs: add 'name=value' decoding for SAS specific lpage
  - examples+utils subdirectories: remove symlinks
  - synchronize man pages with usage messages
  - sg3_utils.spec: rework

Changelog for sg3_utils-1.22 [20061016]
  - sgp_dd: accept verbose=<n> as well as deb=<n> to ease
    interworking with sg_dd and sgm_dd
  - sg_sat_set_features: added to examples directory
  - sg_lib: sync asc/ascq text with spc4r06
    - move SG_LIB_CAT_NO_SENSE and SG_LIB_CAT_RECOVERED to
      20 and 21 respectively; add SG_LIB_CAT_ABORTED_COMMAND
      at 11 (its sense key value) 
  - sg_vpd: tweak '--page=sp --quiet' output
    - change '-HHH' so same as '-rr' (prepares ATA Information
      (ai) response for hdparm)
  - sg_requests: add '-s' option to set exit status from
    parameter data
  - sg_modes: exit quickly from '-e' if device not ready
  - sg_logs: sync sas log pages with sas2r05a
    - expand background scan results log page
    - add '-m=<max_len>' to limit response length
    - drop '-scum' and '-sthr' options and add '-select'
  - sg_write_long: add '--16' option to send 16 byte cdb
    - add '--wr_uncor' and '--pblock' options
  - sg_senddiag: cleanup and add sdiag_sas_p1_stop.txt
    to examples directory
  - sg_format: add '--cmplst=<n>' option (default: 1)
    - add '--pfu=<n>' option
    - expand man page to discuss P/D/C/GLISTs
  - sg_reassign: add '--primary' option to fetch primary
    defect list (PLIST) length
  - sg_readcap: add '-H' option to output response in hex
    and '-r' to output response in binary to stdout
    - add logical blocks per physical block (sbc3r07)
  - sginfo: add PLIST and GLIST designation to defect lists
  - sg_cmds: split this support file into sg_cmds_basic.[hc]
    and sg_cmds_extra.[hc]
    - add sg_ll_ata_pt() (SATL ATA pass) to sg_cmds_extra.[hc]
  - sg_rdac: fix includes for FreeBSD
  - sg_dd: add 'coe_limit=' option to exit after <n>
    consecutive 'coe' type read errors
  - sgm_dd: print out throughput information when signal arrives
    if time=1 (like sg_dd does already)
  - sg_inq: change '-HHH' so same as '-rr'. Now sg_inq, sg_vpd
    and sdparm output for hdparm with '-HHH'
    -add '-l=<resp_len>' option
  - sg_read_long: add '--pblock' option for physical blocks
  - sg_luns: add '--hex' and '--raw' options

src/sg_rem_rest_elem.c
/*
 * Copyright (c) 2022-2023 Douglas Gilbert.
 * All rights reserved.
 * Use of this source code is governed by a BSD-style
 * license that can be found in the BSD_LICENSE file.
 *
 * SPDX-License-Identifier: BSD-2-Clause
 */

#include <unistd.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <stdbool.h>
#include <string.h>
#include <ctype.h>
#include <getopt.h>
#define __STDC_FORMAT_MACROS 1
#include <inttypes.h>

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include "sg_lib.h"
#include "sg_lib_data.h"
#include "sg_pt.h"
#include "sg_cmds_basic.h"
#include "sg_unaligned.h"
#include "sg_pr2serr.h"

/* A utility program originally written for the Linux OS SCSI subsystem.
 *
 * This program issues one of the following SCSI commands:
 *   - REMOVE ELEMENT AND TRUNCATE
 *   - RESTORE ELEMENTS AND REBUILD
 */

static const char * version_str = "1.02 20230503";

#define REMOVE_ELEM_SA 0x18
#define RESTORE_ELEMS_SA 0x19

#define SENSE_BUFF_LEN 64       /* Arbitrary, could be larger */
#define DEF_PT_TIMEOUT  60      /* 60 seconds */


static struct option long_options[] = {
        {"capacity", required_argument, 0, 'c'},
        {"element", required_argument, 0, 'e'},
        {"help", no_argument, 0, 'h'},
        {"quick", no_argument, 0, 'q'},
        {"remove", no_argument, 0, 'r'},
        {"restore", no_argument, 0, 'R'},
        {"timeout", required_argument, 0, 't'},
        {"tmo", required_argument, 0, 't'},
        {"verbose", no_argument, 0, 'v'},
        {"version", no_argument, 0, 'V'},
        {0, 0, 0, 0},
};

static const char * remove_cmd_s = "Remove element and truncate";
static const char * restore_cmd_s = "Restore elements and rebuild";


static void
usage()
{
    pr2serr("Usage: "
            "sg_rem_rest_elem  [--capacity=RC] [--element=EID] [--help] "
            "[--quick]\n"
            "                         [--remove] [--restore] [--timeout=SE] "
            "[--verbose]\n"
            "                         [--version] DEVICE\n");
    pr2serr("  where:\n"
            "    --capacity=RC|-c RC    RC is requested capacity (unit: "
            "block; def: 0)\n"
            "    --element=EID|-e EID    EID is the element identifier to "
            "remove;\n"
            "                            default is 0 which is an invalid "
            "EID\n"
            "    --help|-h          print out usage message\n"
            "    --quick|-q         bypass 15 second warn and wait\n"
            "    --remove|-r        issue REMOVE ELEMENT AND TRUNCATE "
            "command\n"
            "    --restore|-R       issue RESTORE ELEMENTS AND REBUILD "
            "command\n"
            "    --timeout=SE|-t SE    command timeout in seconds (def: "
            "60 secs)\n"
            "    --verbose|-v       increase verbosity\n"
            "    --version|-V       print version string and exit\n\n"
            "Performs a SCSI REMOVE ELEMENT AND TRUNCATE or RESTORE "
            "ELEMENTS AND\nREBUILD command. Either the --remove or "
            "--restore option needs to be given.\n");
}

/* Return of 0 -> success, various SG_LIB_CAT_* positive values or -1 ->
 * other errors */
static int
sg_ll_rem_rest_elem(int sg_fd, int sa, uint64_t req_cap, uint32_t e_id,
                    int tmo, bool noisy, int verbose)
{
    int ret, res, sense_cat;
    struct sg_pt_base * ptvp;
    uint8_t sai16_cdb[16] =
          {SG_SERVICE_ACTION_IN_16, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
           0, 0, 0, 0};
    uint8_t sense_b[SENSE_BUFF_LEN] SG_C_CPP_ZERO_INIT;
    const char * cmd_name;

    sai16_cdb[1] = 0x1f & sa;
    if (REMOVE_ELEM_SA == sa) {
        sg_put_unaligned_be64(req_cap, sai16_cdb + 2);
        sg_put_unaligned_be32(e_id, sai16_cdb + 10);
        cmd_name = remove_cmd_s;
    } else
        cmd_name = restore_cmd_s;
    if (verbose) {
        char d[128];

        pr2serr("    %s cdb: %s\n", cmd_name,
                sg_get_command_str(sai16_cdb, 16, false, sizeof(d), d));
    }

    ptvp = construct_scsi_pt_obj();
    if (NULL == ptvp) {
        pr2serr("%s: out of memory\n", cmd_name);
        return -1;
    }
    set_scsi_pt_cdb(ptvp, sai16_cdb, sizeof(sai16_cdb));
    set_scsi_pt_sense(ptvp, sense_b, sizeof(sense_b));
    res = do_scsi_pt(ptvp, sg_fd, tmo, verbose);
    ret = sg_cmds_process_resp(ptvp, cmd_name, res, noisy,
                               verbose, &sense_cat);
    if (-1 == ret) {
        if (get_scsi_pt_transport_err(ptvp))
            ret = SG_LIB_TRANSPORT_ERROR;
        else
            ret = sg_convert_errno(get_scsi_pt_os_err(ptvp));
    } else if (-2 == ret) {
        switch (sense_cat) {
        case SG_LIB_CAT_RECOVERED:
        case SG_LIB_CAT_NO_SENSE:
            ret = 0;
            break;
        default:
            ret = sense_cat;
            break;
        }
    } else
        ret = 0;
    destruct_scsi_pt_obj(ptvp);
    return ret;
}


int
main(int argc, char * argv[])
{
    bool quick = false;
    bool reat = false;
    bool resar = false;
    bool verbose_given = false;
    bool version_given = false;
    int res, c;
    int sg_fd = -1;
    int tmo = DEF_PT_TIMEOUT;
    int verbose = 0;
    int ret = 0;
    int sa = 0;
    uint32_t e_id = 0;
    uint64_t req_cap = 0;
    int64_t ll;
    const char * device_name = NULL;
    const char * cmd_name;

    while (1) {
        int option_index = 0;

        c = getopt_long(argc, argv, "c:e:hqrRt:vV", long_options,
                        &option_index);
        if (c == -1)
            break;

        switch (c) {
        case 'c':
            ll = sg_get_llnum(optarg);
            if (-1 == ll) {
                pr2serr("--capacity= expects a numeric argument\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            req_cap = (uint64_t)ll;
            break;
        case 'e':
            ll = sg_get_llnum(optarg);
            if ((ll < 0) || (ll > UINT32_MAX)) {
                pr2serr("bad argument to '--element=EID'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            if (0 == ll)
                pr2serr("Warning: 0 is an invalid element identifier\n");
            e_id = (uint64_t)ll;
            break;
        case 'h':
        case '?':
            usage();
            return 0;
        case 'q':
            quick = true;
            break;
        case 'r':
            reat = true;
            sa = REMOVE_ELEM_SA;
            break;
        case 'R':
            resar = true;
            sa = RESTORE_ELEMS_SA;
            break;
        case 't':
            tmo = sg_get_num(optarg);
            if (tmo < 0) {
                pr2serr("bad argument to '--timeout=SE'\n");
                return SG_LIB_SYNTAX_ERROR;
            }
            break;
        case 'v':
            verbose_given = true;
            ++verbose;
            break;
        case 'V':
            version_given = true;
            break;
        default:
            pr2serr("unrecognised option code 0x%x ??\n", c);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }
    if (optind < argc) {
        if (NULL == device_name) {
            device_name = argv[optind];
            ++optind;
        }
        if (optind < argc) {
            for (; optind < argc; ++optind)
                pr2serr("Unexpected extra argument: %s\n",
                        argv[optind]);
            usage();
            return SG_LIB_SYNTAX_ERROR;
        }
    }

#ifdef DEBUG
    pr2serr("In DEBUG mode, ");
    if (verbose_given && version_given) {
        pr2serr("but override: '-vV' given, zero verbose and continue\n");
        verbose_given = false;
        version_given = false;
        verbose = 0;
    } else if (! verbose_given) {
        pr2serr("set '-vv'\n");
        verbose = 2;
    } else
        pr2serr("keep verbose=%d\n", verbose);
#else
    if (verbose_given && version_given)
        pr2serr("Not in DEBUG mode, so '-vV' has no special action\n");
#endif
    if (version_given) {
        pr2serr("version: %s\n", version_str);
        return 0;
    }

    if (1 != ((int)reat + (int)resar)) {
        pr2serr("One, and only one, of these options needs to be given:\n"
                "   --remove or --restore\n\n");
        usage();
        return SG_LIB_CONTRADICT;
    }
    cmd_name = reat ? remove_cmd_s : restore_cmd_s;

    if (0 == tmo)
        tmo = DEF_PT_TIMEOUT;
    if (NULL == device_name) {
        pr2serr("missing device name!\n");
        usage();
        return SG_LIB_SYNTAX_ERROR;
    }

    sg_fd = sg_cmds_open_device(device_name, false /* rw */, verbose);
    if (sg_fd < 0) {
        int err = -sg_fd;
        if (verbose)
            pr2serr("open error: %s: %s\n", device_name,
                    safe_strerror(err));
        ret = sg_convert_errno(err);
        goto fini;
    }
    if (! quick) {
        int k;
        char b[80] SG_C_CPP_ZERO_INIT;
        char ch;

        for (k = 0; k < (int)sizeof(b) - 1; ++k) {
            ch = cmd_name[k];
            if ('\0' == ch)
                break;
            else if (islower(ch))
                b[k] = toupper(ch);
            else
                b[k] = ch;
        }
        sg_warn_and_wait(b, device_name, false);
    }

    res = sg_ll_rem_rest_elem(sg_fd, sa, req_cap, e_id, tmo, true, verbose);
    ret = res;
    if (res) {
        if (SG_LIB_CAT_INVALID_OP == res)
            pr2serr("%s command not supported\n", cmd_name);
        else {
            char b[80];

            sg_get_category_sense_str(res, sizeof(b), b, verbose);
            pr2serr("%s command: %s\n", cmd_name, b);
        }
    }

fini:
    if (sg_fd >= 0) {
        res = sg_cmds_close_device(sg_fd);
        if (res < 0) {
            pr2serr("close error: %s\n", safe_strerror(-res));
            if (0 == ret)
                ret = sg_convert_errno(-res);
        }
    }
    if (0 == verbose) {
        if (! sg_if_can2stderr("sg_rem_rest_elem failed: ", ret))
            pr2serr("Some error occurred, try again with '-v' or '-vv' for "
                    "more information\n");
    }
    return (ret >= 0) ? ret : SG_LIB_CAT_OTHER;
}
